blog/02-15-2023-build-feature-rich-persistent-dapp-on-celo-using-wagmi/build-feature-rich-persistent-dapp-on-celo-using-wagmi.md

```diff
@@ -0,0 +1,747 @@
+
+```js
+contract SwapLab is Ownable, ReentrancyGuard {
+
+  function swapERC20ForCelo(address asset) payable public nonReentrant {
+    SupportedAsset memory sat = supportedAssets[asset];
+    if(!sat.isSupported) revert UnSupportedAsset(asset);
+    uint mantissa = 10 ** IERC20Metadata(asset).decimals();
+    if(Address.isContract(msg.sender)) revert ContractAddressNotAllowed(msg.sender);
+    uint amountToSwap = IERC20(asset).allowance(msg.sender, address(this));
+    require(amountToSwap >= sat.minimumSwapValue, "Insufficient allowance");
+    uint amountCeloToReceive = amountToSwap.mul(sat.rate).div(mantissa).sub(sat.swapFee);
+    uint bal = address(this).balance;
+    if(amountCeloToReceive > bal) revert InsufficientFund(bal, amountCeloToReceive);
+    require(IERC20(asset).transferFrom(msg.sender, owner(), amountToSwap), "Failed");
+    (bool sent, ) = address(msg.sender).call{value: amountCeloToReceive}('');
+    require(sent, 'Anomally detected');
+  }
+
+  function deposit() payable public {
+    require(msg.value > 0,"Thank you");
+  }
+
+  function withdraw(address to, uint amount) public onlyOwner {
+    require(to != address(0), 'Zero recipient');
+    (bool done,) = to.call{value: amount}('');
+    require(done, 'Failed');
+  }
+}
+```
+
+**Compile**
+
+In the `backend` directory, run: 
+
+```bash
+npx hardhat compile
+```
+
+![image](images/3.png)
+
+
+**Test**
+
+```bash
+npx hardhat test
+```
+![image](images/5.png)
+
+**Deploy**
+
+```bash
+yarn deploy
+```
+
+When you run the deploy command,  a folder named `deployments` is created where the deployment information is saved. There we'll import the 'Abi' and 'contract address' to the frontend where we'll need them.
+
+
+## Frontend
+
+Exit the current directory into the root folder and install the dependencies: 
+
+```bash
+cd ..
+cd frontend
+yarn install
+```
+
+Since I already completed the setup, it's important that I show you how to set up the frontend folder.
+
+If we are to create a new project from the scratch, let's assume we're in the root directory with the `backend` being the only folder. The recommended way to start a wagmi project is to run the following command.
+
+```bash
+npm init wagmi -- --template next-connectkit
+```
+Select the choice of provider.
+
+![image](images/2.png)
+
+Installation success
+
+![image](images/6.png)
+
+The command takes extra arguments `--template` and `next-connectkit` to start the project with a boilerplate that uses connectkit compatible with NextJs to display information to users on the frontend.
+
+
+Follow the instructions to make your choice. For this project, I have selected the 'default' option which uses 'NextJs'. You really do not need to do anything else. Much of the work is done for you in the background.
```

Comment:
Follow the instructions to choose your choice. For this project, I have selected the 'default' option, using 'NextJs'. You do not need to do anything else. Much of the work is done in the background.